### PR TITLE
feat(dht_node): add request and responses for pings

### PR DESCRIPTION
**Pull request by zetok**
This change is [Reviewable](https://reviewable.io/reviews/zetok/tox/71)

### CODE REVIEW

File: src/toxcore/dht_node.rs [hunk 1]
```diff
@@ -95,6 +95,56 @@ impl DhtNode {
...
+        let shared_secret = &encrypt_precompute(peer_pk, &self.dht_secret_key);
+        let nonce = &gen_nonce();
+        DhtPacket::new(shared_secret,
+                       &self.dht_public_key,
+                       nonce,
+                       &ping)
+    }
+
+    /**
+    Request nodes from a peer. Peer might or might not even reply.
```
Comment (kpp) [inline src/toxcore/dht_node.rs]: Request nodes?

File: src/toxcore/dht_node.rs [hunk 2]
```diff
@@ -95,6 +95,56 @@ impl DhtNode {
...
     }
 
+
+    /**
+    Create a [`DhtPacket`](../dht/struct.DhtPacket.html) to peer with `peer_pk`
+    `PublicKey` containing a [`PingReq`](../dht/struct.PingReq.html) request.
+    */
+    fn create_ping_req(&self, peer_pk: &PublicKey) -> DhtPacket {
+        let ping = PingReq::new();
+        let shared_secret = &encrypt_precompute(peer_pk, &self.dht_secret_key);
```
Comment (kpp) [inline src/toxcore/dht_node.rs]: Add TODO: `precompute shared key to calc it 1 time`

File: src/toxcore/dht_node.rs [hunk 3]
```diff
@@ -95,6 +95,56 @@ impl DhtNode {
...
+        DhtPacket::new(shared_secret,
+                       &self.dht_public_key,
+                       nonce,
+                       &ping)
+    }
+
+    /**
+    Request nodes from a peer. Peer might or might not even reply.
+
+    Creates a future for sending request for nodes.
```
Comment (kpp) [inline src/toxcore/dht_node.rs]: `sending request`?

File: src/toxcore/dht_node.rs [hunk 4]
```diff
@@ -95,6 +95,56 @@ impl DhtNode {
...
+    [`DhtPacket`]: ../dht/struct.DhtPacket.html
+    [ping response]: ../dht/struct.PingResp.html
+    */
+    pub fn respond_ping(&self,
+                        sink: UdpFramed<ToxCodec>,
+                        peer: &PackedNode,
+                        request: &DhtPacket)
+        -> Option<sink::Send<UdpFramed<ToxCodec>>>
+    {
+        let precomp = encrypt_precompute(peer.pk(), &self.dht_secret_key);
```
Comment (kpp) [inline src/toxcore/dht_node.rs]: add todo

File: src/toxcore/dht_node.rs (hunk 1 above)
Comment (zetok) [reply src/toxcore/dht_node.rs]: Fixed.

File: src/toxcore/dht_node.rs [hunk 5]
```diff
@@ -287,141 +343,292 @@ mod test {
...
+        handle.spawn(bob_request.then(|_| ok(())));
+
+        let received = core.run(future_recv).unwrap();
+        let (_alice_socket, recv_buf, size, _saddr) = received;
+        assert!(size != 0);
+
+        let recv_packet = DhtPacket::from_bytes(&recv_buf[..size]).unwrap();
+        let payload: PingReq = recv_packet
+            .get_packet(&alice.dht_secret_key)
+            .expect("Failed to decrypt payload");
```
Comment (kpp) [inline src/toxcore/dht_node.rs:416]: + add case for Eve

File: src/toxcore/dht_node.rs [hunk 6]
```diff
@@ -287,141 +343,292 @@ mod test {
...
+        // eve can't decrypt the request
+        let alice = DhtNode::new().unwrap();
+        let (bob_pk, bob_sk) = gen_keypair();
+        let (_, eve_sk) = gen_keypair();
+        let packet1 = alice.create_getn(&bob_pk);
+        assert_eq!(*alice.dht_public_key, packet1.sender_pk);
         assert_eq!(PacketKind::GetN, packet1.kind());
 
-        let payload1: GetNodes = packet1.get_packet(&peer_sk)
+        // eve can't even
```
Comment (kpp) [inline src/toxcore/dht_node.rs]: `can't even` what?

File: src/toxcore/dht_node.rs (hunk 5 above)
Comment (zetok) [reply src/toxcore/dht_node.rs:416]: This is checked in test above, which tests function that deals with crypto, should the check still be added here?

File: src/toxcore/dht_node.rs (hunk 2 above)
Comment (zetok) [reply src/toxcore/dht_node.rs]: done

File: src/toxcore/dht_node.rs (hunk 4 above)
Comment (zetok) [reply src/toxcore/dht_node.rs]: done

File: src/toxcore/dht_node.rs (hunk 6 above)
Comment (zetok) [reply src/toxcore/dht_node.rs]: fixed

File: src/toxcore/dht_node.rs (hunk 5 above)
Comment (kpp) [reply src/toxcore/dht_node.rs:416]: no. you may add get_packet<DIFFERENT_TYPE> to check that the payload cannot be parsed.

Comment (zetok) [reply src/toxcore/dht_node.rs:416]: Done for the packets that can't be parsed as other packets.

**Comment by kpp** [comment]
:lgtm:

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zetok/tox/71#-:-KuEGAJB41RBJT6DrhVJ:bnfp4nl)*

**Pull request merged**